// formdata-node/lib/cjs/util/deprecations.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deprecateSymbolAsyncIterator = exports.deprecateGetComputedLength = exports.deprecateOptions = exports.deprecateReadStream = exports.deprecateBuffer = exports.deprecateStream = exports.deprecateHeaders = exports.deprecateBoundary = void 0;
const util_1 = require("util");
const noop_1 = __importDefault(require("./noop"));
exports.deprecateBoundary = util_1.deprecate(noop_1.default, "FormData#boundary property is non-standard and will be removed "
    + "from this package in the next major release (4.x). "
    + "Use https://npmjs.com/form-data-encoder package to serilize FormData.");
exports.deprecateHeaders = util_1.deprecate(noop_1.default, "FormData#headers property is non-standard and will be removed "
    + "from this package in the next major release (4.x). "
    + "Use https://npmjs.com/form-data-encoder package to serilize FormData.");
exports.deprecateStream = util_1.deprecate(noop_1.default, "FormData#stream property is non-standard and will be removed "
    + "from this package in the next major release (4.x). "
    + "Use https://npmjs.com/form-data-encoder package to serilize FormData.");
exports.deprecateBuffer = util_1.deprecate(noop_1.default, "The usage of Buffer in entry's value is deprecated. "
    + "Use Blob, File, fileFromPath() or fileFromPathSync() instead.");
exports.deprecateReadStream = util_1.deprecate(noop_1.default, "The usage of ReadStream in entry's value is deprecated. "
    + "Use Blob, File, fileFromPath or fileFromPathSync() instead.");
exports.deprecateOptions = util_1.deprecate(noop_1.default, "The options argument is non-standard and will be removed "
    + "from this package in the next major release (4.x).");
exports.deprecateGetComputedLength = util_1.deprecate(noop_1.default, "FormData#getComputedLength() method is non-standard and will be "
    + "removed from this package in the next major release (4.x). "
    + "Use https://npmjs.com/form-data-encoder package to serilize FormData.");
exports.deprecateSymbolAsyncIterator = util_1.deprecate(noop_1.default, "FormData#[Symbol.asyncIterator]() method is non-standard and will be "
    + "removed from this package in the next major release (4.x). "
    + "Use https://npmjs.com/form-data-encoder package to serilize FormData.");
